File: maximum_twin_sum_of_a_linked_list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode second = null;

        while (slow != null) {
            ListNode temp = slow.next;
            slow.next = second;
            second = slow;
            slow = temp;
        }

        int res = 0;
        while (second != null) {
            res = Math.max(res, head.val + second.val);
            head = head.next;
            second = second.next;
        }

        return res;        
    }
}

File: maximum_twin_sum_of_a_linked_list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;

        while (fast != nullptr && fast->next != nullptr) {
            slow = slow->next;
            fast = fast->next->next;
        }

        ListNode* second = nullptr;

        while (slow != nullptr) {
            ListNode* temp = slow->next;
            slow->next = second;
            second = slow;
            slow = temp;
        }

        int res = 0;
        while (second != nullptr) {
            res = std::max(res, head->val + second->val);
            head = head->next;
            second = second->next;
        }

        return res;        
    }
};
